### ISSUE TITLE
After song is over, there often some stuck notes playing on some drives

### ISSUE THREAD

**Issue by ostrosablin** [samphonic/MoppyClassic#86]
It appears that too often, after song is over (or if you click stop), some stuck notes are left that play indefinitely. I guess that happens if there is some note that is played in the end of song and Moppy will never send stop signal for that channel, they could be only stopped by clicking Stop/Reset. So, I propose several changes:

1) Adding auto-reset-in-the-end checkbox would be nice, because silencing floppies manually is not very convenient, especially if you're recording them on video.
2) Perhaps, it's better to reset the song position slider when Moppy hits end of the song? It's logical, because if song has ended, you couldn't play it from this position anyway.
3) For some reason, clicking stop/reset after song has ended wouldn't rewind it. If 2 doesn't sounds good, then rewind could be set to stop/reset button.

That would clearly improve Moppy's usability, especially when playing MIDI songs that you didn't originally create.

**Comment by samphonic** [comment]
The stuck notes are usually only an issue with drive pooling (which I think you mentioned in another ticket that you were using).  Drive pooling is pretty rough around the edges and was more of a quick stop-gap measure to help with playing the drives in real time from a keyboard.  It's probably possible to adjust the pooling strategy, but it might take some work.  In the meantime, avoiding MIDIs with more simultaneous notes than output pooled drives should prevent the issue from occurring (I think).

As for your suggestions:
1. The checkbox is a good idea.  I didn't reset the drives automatically at the end because when recording them on video, I wanted to avoid ending every song with a big _FRRNT_.  What I might do though is just make a call to stop all the current notes when the sequence ends so that at least there aren't any notes hanging over.
2. Yeah, might add that eventually.
3. Clicking the stop/reset button a second time should (if I remember correctly) reset the song back to the beginning.  It's more of a "pause/reset" button.

Not sure when I'll get around to modifying any code for non-show-stopping issues, but I'll hang on to this issue for when I do.

**Comment by ostrosablin** [comment]
Yeah, sending forced note-off sounds great, even better than auto-resetting, because it won't cause head parking noise. However, in regards to 3 - no, second click won't reset song to beginning. It will reset only if song is playing now, so it's "stop/reset" after all. If it's paused or reached end - it will only reset drives.

Perhaps I'll even be able to contribute to the project (since I could code in Java just fine, and those changes shouldn't be hard to pull off) and submit them as pull request, since you mostly approved them.

**Comment by tmp6154** [comment]
I've already started. I've implemented 2 like that

```java
    private void updateProgressDisplay() {
        long currentSeconds = seq.getSecondsPosition();
        long totalSeconds = seq.getSecondsLength();
        if ((!seq.isRunning()) && (currentSeconds == totalSeconds)) {
        	seq.setSecondsPosition(0);
        }
```

**Comment by ostrosablin** [comment]
I've already started. I thought whether to implement 2 in sequencer or in UI code. I've implemented 2 in UI like that.

``` java
    private void updateProgressDisplay() {
        long currentSeconds = seq.getSecondsPosition();
        long totalSeconds = seq.getSecondsLength();
        if ((!seq.isRunning()) && (currentSeconds == totalSeconds)) {
            currentSeconds = 0;
            seq.resetSequencer();
        }
```

Should work well and not break anything. If sequencer has stopped and song is at the end - we rewind the song (and slider will rewind as well). Is it fine, or should it go in sequencer code?

Also, added call to resetSequencer for reset, now second click will actually reset song to the beginning.

**Comment by ostrosablin** [comment]
Implemented an autoreset checkbox just now. It works perfectly. Now all that remains is to force note-off sending for pause, stop and normal song ending. Probably most difficult part, because I haven't figured out how to do that yet.

**Comment by samphonic** [comment]
The MoppyCOMBridge class has a silenceDrives() method that will do exactly what you're looking for.  You may not have direct access to the MoppyCOMBridge from where the controls are happening though, and may have to add methods to some intermediate classes.

**Comment by ostrosablin** [comment]
Well, thanks for the hint, I was able to do it. I ended up adding a second timer with low delay (50 ms) into SequencerControls, dedicated to checking for sequencer stop (because it looks like Sequencer has no way to set up a stop event notification, which would've be more appropriate). ProgressTimer was not fast enough and it still produced noticeable amount of stuck notes in end.

Now that I've implemented all suggestions, I'll make a pull request.
